pkg/ioutils/fswriters.go

```diff
@@ -17,6 +19,9 @@ type AtomicFileWriterOptions struct {
 	// On successful return from Close() this is set to the mtime of the
 	// newly written file.
 	ModTime time.Time
+	// Whenever an atomic file is successfully written create a temporary
+	// file of the same size in order to pre-allocate storage for the next write
+	PreAllocate bool
```

Comment:
The `PreAllocate` and `!PreAllocate` code paths differ _so much_, and very importantly they have different security implications (`PreAllocate` must not be used in directories with hostile writers, like `*/tmp`), that I don’t think it makes much sense for them to use the same object at all.

The comparatively small parts of the code that can be shared between the two can be shared another way (e.g. using a shared helper function).

Reply:
> `PreAllocate` must not be used in directories with hostile writers

Not just that, `PreAllocate` is also _only_ correct if there is some external-to-`pkg/ioutils` locking that ensures serialization of uses of `AtomicFileWriter`. (That’s fine for the container/image/layer stores, we have locks there. But it’s a new responsibility for callers of this code, and it needs to be very clearly documented.)